### DISCUSSION TITLE
Encoder API inconsistency

### DISCUSSION THREAD

**Discussion by verdverm** [cue-lang/cue#2116] [Q&A]
Noticed that JSON & Yaml have different signatures for their `Extract` functions, returning `ast.Expr` vs `ast.File` respectively.

- https://pkg.go.dev/cuelang.org/go@v0.4.3/encoding/json#Extract
- https://pkg.go.dev/cuelang.org/go@v0.4.3/encoding/yaml#Extract

**Comment by myitcv** [comment]
@verdverm is there a specific issue with these APIs being inconsistent?

**Reply by verdverm** [reply]
If you want to have a processing pipeline (i.e. adding data files from CLI args to the build instance), you have to create different workflows for the different formats. For JSON, you have to additionally create an `ast.File`, add the filename, and also lift the fields from the unnamed top-level json struct.

Like this: https://github.com/hofstadter-io/hof/blob/_dev/lib/cuetils/runtime.go#L140
